### PR TITLE
Breadcrumbs on background thread

### PR DESCRIPTION
**Pull request by victorhurdugaci**
@ChengTian @glennc @davidfowl

### CODE REVIEW

File: src/Microsoft.Framework.Runtime/DependencyManagement/NuGetDependencyResolver.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 using System.Runtime.Versioning;
 using Microsoft.Framework.Runtime.DependencyManagement;
 using NuGet;
+using System.Threading;
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/DependencyManagement/NuGetDependencyResolver.cs]: This is in the wrong place

**Comment by victorhurdugaci** [comment]
Forgot to mention, we save ~100ms at startup with the background thread

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 2]
```diff
@@ -109,7 +133,43 @@ private void CreateBreadcrumbFile(string fileName)
...
         {
-            Logger.TraceInformation("Failed to write servicing breadcrumb for {0} because an exception was thrown: {1}", fileName, exception);
+            Logger.TraceInformation(
+                "[{0}] Failed to write servicing breadcrumb for {1} because an exception was thrown: {2}",
+                _logType,
+                fileName,
+                exception);
+        }
+
+        private static BreadcrumbInfo GetRuntimeBreadcrumb()
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Can we do this somewhere else?

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 3]
```diff
@@ -5,12 +5,18 @@
...
 
 namespace Microsoft.Framework.Runtime.Servicing
 {
     public class Breadcrumbs
     {
+        private static readonly string _logType = typeof(Breadcrumbs).GetType().Name;
         private readonly string _breadcrumbsFolder;
+        private readonly Queue<BreadcrumbInfo> _breadcrumbsQueue = new Queue<BreadcrumbInfo>();
+
+        public static Breadcrumbs Current { get; private set; } = new Breadcrumbs();
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Current makes it sounds like there's more than one. Get rid of this altogether.

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 4]
```diff
@@ -5,12 +5,18 @@
...
+using System.Collections.Generic;
+using System.Threading;
 
 namespace Microsoft.Framework.Runtime.Servicing
 {
     public class Breadcrumbs
     {
+        private static readonly string _logType = typeof(Breadcrumbs).GetType().Name;
         private readonly string _breadcrumbsFolder;
+        private readonly Queue<BreadcrumbInfo> _breadcrumbsQueue = new Queue<BreadcrumbInfo>();
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Just use a list.

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 5]
```diff
@@ -37,36 +46,48 @@ public static bool IsPackageServiceable(PackageInfo package)
             return false;
         }
 
-        public void CreateBreadcrumb(string packageId, SemanticVersion packageVersion)
+        public void EnqueueBreadcrumb(string packageId, SemanticVersion packageVersion)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: AddBreadcrumb?

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 6]
```diff
@@ -37,36 +46,48 @@ public static bool IsPackageServiceable(PackageInfo package)
...
+        }
 
-            if (!string.IsNullOrWhiteSpace(version))
+        /// <summary>
+        /// Writes all the enqueued breadcrumbs to disk. 
+        /// </summary>
+        /// <returns></returns>
+        public void ProcessBreadcrumbsQueue()
+        {
+            while (_breadcrumbsQueue.Count > 0)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Just loop over the list and mark each one. The clear it.

File: src/Microsoft.Framework.Runtime/DependencyManagement/NuGetDependencyResolver.cs [hunk 7]
```diff
@@ -272,6 +272,16 @@ public void Initialize(IEnumerable<LibraryDescription> packages, FrameworkName t
 
                 dependency.LoadableAssemblies = assemblies;
             }
+
+            // Process the breadcrumbs on a different thread
+            // but don't wait for it
+            var breadcrumbsThread = new Thread(() =>
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/DependencyManagement/NuGetDependencyResolver.cs]: Use the thread pool instead of making a dedicated thread.

**Comment by victorhurdugaci** [comment]
@davidfowl updated the PR. What do you think?

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 8]
```diff
@@ -109,7 +139,17 @@ private void CreateBreadcrumbFile(string fileName)
 
         private static void LogBreadcrumbsCreationFailure(string fileName, Exception exception)
         {
-            Logger.TraceInformation("Failed to write servicing breadcrumb for {0} because an exception was thrown: {1}", fileName, exception);
+            Logger.TraceInformation(
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: TraceError

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 9]
```diff
@@ -37,36 +45,55 @@ public static bool IsPackageServiceable(PackageInfo package)
...
+        }
 
-            // Create both files for now until we get clear instructions about the format of the name
-            CreateBreadcrumbFile(packageId);
-            CreateBreadcrumbFile(packageId + "." + packageVersion);
+        public void WriteAllBreadcrumbs(bool background = false)
+        {
+            if (background)
+            {
+                ThreadPool.QueueUserWorkItem(_ =>
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: If we actually want the process to wait until breadcrumbs are written then we need to use a normal thread :cry:

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Add an environment variable that makes this NOOP. I want to disable this when I'm running performance tests.

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Leave the ThreadPool.QueueUserWorkItem it's fine.

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 10]
```diff
@@ -2,15 +2,25 @@
...
-using System.Reflection;
+using System.Threading;
 using NuGet;
 
 namespace Microsoft.Framework.Runtime.Servicing
 {
     public class Breadcrumbs
     {
+        // It is recommended that you disable breadcrumbs only for perf runs
+        private const string NoBreadcrumbsEnvironmentVariableName = "DNX_NO_BREADCRUMBS";
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: @troydai you might want to set this env var for perf runs

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs (hunk 9 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Added the NOOP variable

**Comment by davidfowl** [comment]
What happens when there is no servicing folder? None of this code runs right? It shoud all noop in that case. 

Did you test this out on mono?

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs (hunk 10 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Thanks for reminding.

**Comment by victorhurdugaci** [comment]
@davidfowl No service folder = noop. Mono = tested already (no folder there)

**Comment by davidfowl** [comment]
Where exactly does it noop?

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 11]
```diff
@@ -27,46 +46,83 @@ public Breadcrumbs(string breadcrumbsFolder)
...
+        /// <summary>
+        /// Writes a breadcrumb on the disk 
+        /// </summary>
+        /// <param name="packageId">The ID of the package</param>
+        /// <param name="packageVersion">The version of the the package</param>
+        private void CreateBreadcrumb(string packageId, SemanticVersion packageVersion)
+        {
+            if (string.IsNullOrWhiteSpace(_breadcrumbsFolder))
+            {
+                return;
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: @davidfowl "Where exactly does it noop?" Here. I could change it to use the `_isEnabled` property instead an really do nothing.

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: That was one of those leading questions. Use that flag to represent both cases.

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: @davidfowl updated

**Comment by victorhurdugaci** [comment]
Any more comments/concerns on this?

**Pull request merged**